server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -298,22 +305,173 @@ private CoordinationDiagnosticsResult diagnoseOnHaveNotSeenMasterRecently(Master
         Collection<DiscoveryNode> masterEligibleNodes = getMasterEligibleNodes();
         final CoordinationDiagnosticsResult result;
         boolean leaderHasBeenElected = coordinator.getPeerFinder().getLeader().isPresent();
-        if (masterEligibleNodes.isEmpty() && leaderHasBeenElected == false) {
+        boolean noLeaderElectedAndNoMasterEligibleNodes = leaderHasBeenElected == false && masterEligibleNodes.isEmpty();
+        boolean localNodeIsMasterEligible = clusterService.localNode().isMasterNode();
+        if (noLeaderElectedAndNoMasterEligibleNodes) {
             result = getResultOnNoMasterEligibleNodes(localMasterHistory, explain);
         } else if (leaderHasBeenElected) {
             DiscoveryNode currentMaster = coordinator.getPeerFinder().getLeader().get();
             result = getResultOnCannotJoinLeader(localMasterHistory, currentMaster, explain);
-        } else {
+        } else if (localNodeIsMasterEligible == false) { // none is elected master and we aren't master eligible
             // NOTE: The logic in this block will be implemented in a future PR
             result = new CoordinationDiagnosticsResult(
                 CoordinationDiagnosticsStatus.RED,
                 "No master has been observed recently",
                 CoordinationDiagnosticsDetails.EMPTY
             );
+        } else { // none is elected master and we are master eligible
+            result = diagnoseOnHaveNotSeenMasterRecentlyAndWeAreMasterEligible(localMasterHistory, masterEligibleNodes, explain);
+        }
+        return result;
+    }
+
+    private CoordinationDiagnosticsResult diagnoseOnHaveNotSeenMasterRecentlyAndWeAreMasterEligible(
```

Comment:
OK it is now static, non-private, and unit tested. I actually kind of like the name (despite its length) because it tells exactly what it is doing. This is one of the deepest leaves in the diagram we're working off of, so it's not surprising that it takes a few words to get there. Have any ideas on how to express it more succinctly without losing the meaning?